cpp/src/sampling/detail/check_edge_bias_values.cuh

```diff
@@ -0,0 +1,73 @@
+/*
+ * Copyright (c) 2024, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include "prims/count_if_e.cuh"
+
+#include <cugraph/edge_property.hpp>
+#include <cugraph/edge_src_dst_property.hpp>
+#include <cugraph/graph_functions.hpp>
+#include <cugraph/graph_view.hpp>
+#include <cugraph/utilities/host_scalar_comm.hpp>
+
+#include <raft/core/handle.hpp>
+
+#include <rmm/device_uvector.hpp>
+
+#include <thrust/count.h>
+#include <thrust/tuple.h>
+
+#include <tuple>
+
+namespace cugraph {
+namespace detail {
+
+template <typename vertex_t, typename edge_t, typename bias_t, bool multi_gpu>
+std::tuple<size_t, size_t> check_edge_bias_values(
+  raft::handle_t const& handle,
+  graph_view_t<vertex_t, edge_t, false, multi_gpu> const& graph_view,
+  edge_property_view_t<edge_t, bias_t const*> edge_bias_view)
+{
+  auto num_negative_edge_weights =
+    count_if_e(handle,
+               graph_view,
+               edge_src_dummy_property_t{}.view(),
+               edge_dst_dummy_property_t{}.view(),
+               edge_bias_view,
+               [] __device__(vertex_t, vertex_t, auto, auto, bias_t b) { return b < 0.0; });
+
+  size_t num_overflows{0};
+  {
+    auto bias_sums = compute_out_weight_sums(handle, graph_view, edge_bias_view);
+    num_overflows  = thrust::count_if(
+      handle.get_thrust_policy(), bias_sums.begin(), bias_sums.end(), [] __device__(auto sum) {
+        return sum > std::numeric_limits<bias_t>::max();
```

Comment:
This check seems suspect to me.  `sum` is going to be of type `bias_t`.  I don't see how it could ever have a value that is strictly greater than `std::numeric_limits<bias_t>::max()`.

Should this check for `std::isinf(sum)` or equality to `std::numeric_limits<bias_t>::max()`?

Reply:
If overflows, it will become std::numeric_limits<bias_t>::infinity();

Reply:
Here's an example...
```
#include <iostream>
#include <limits>

int main(int argc, char **argv) {
  double x = std::numeric_limits<double>::max();
  double y = std::numeric_limits<double>::infinity();
  x += 1;

  if (x > std::numeric_limits<double>::max())
    std::cout << "x comparison works..." << std::endl;
  else
    std::cout << "x comparison fails" << std::endl;

  if (y > std::numeric_limits<double>::max())
    std::cout << "y comparison works..." << std::endl;
  else
    std::cout << "y comparison fails" << std::endl;
}
```

Output when I run this is:
```
x comparison fails
y comparison works...
```

I'm not sure that overflow guarantees that it gets set to infinity.  Of course that's a numerical instability issue more than anything else, so perhaps not worthy of struggling with.

Reply:
Oh, max() + 1 is still max() as max() for double is 1.79769e+308 and # significant digits for double is 16-ish. Adding 1 won't affect any numbers within the significant digit.

And AFAIK, how floating point numbers handle overflow is not strictly defined, but this is what I got from the following code.

```
#include <iostream>
#include <limits>

int main() {
  auto val = std::numeric_limits<double>::max();
  std:: cout << "val=" << val << " val + 1e308=" << (val + 1e308) << " (val + 1e308) - val=" << ((val + 1e308) - val) << " val * 2=" << val * 2 << " val + val=" << (val + val) << std::endl;
  return 0;
}
```

```
val=1.79769e+308 val + 1e308=inf (val + 1e308) - val=inf val * 2=inf val + val=inf
```

Reply:
If I add 1e291, I don't get inf. If I add 1e292, I get inf. This test isn't perfect as floating point arithmetic gets highly inaccurate around their representation limit, but this test detects at least some overflow events.